---
title: "Genomic prediction for quantitative traits"
authors: "E. Tourrette, T. Flutre, B. Servin"
format:
  html:
    toc: true
    toc-depth: 2
    toc-location: left
    number-sections: true
editor: visual
---


# Preamble

Set the paths:
```{r}
projectName <- "QLife2025"
projectDir <- ""
if(Sys.info()["user"] == "tflutre"){
  projectDir <- file.path("~/Bureau/quick/", projectName)
} else if(Sys.info()["user"] == "eliset"){
  projectDir <- file.path("~/Desktop/qlife_2025")
}
projectDir <- path.expand(projectDir)
stopifnot(dir.exists(projectDir))
```

Load the dependencies:
```{r}
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(kinship2))
suppressPackageStartupMessages(library(sommer))
suppressPackageStartupMessages(library(rrBLUP))
```



# Main objective

The goal of this session is to show how to answer the question "how to replace pedigree-based kinship with SNP-based kinship?".



# Run the simulation

```{r}
## TODO
```



# Load all the data

## Pedigree

```{r}
inF <- file.path(projectDir, "simulations", "example_data", "pedigree.txt")
stopifnot(file.exists(inF))
pedDat <- read.table(inF, header=FALSE, sep=" ", stringsAsFactors=FALSE,
                     col.names=c("ind_id", "father_id", "mother_id", "sex",
                                 "affection", "generation",
                                 "phenotype", "breeding_value"))
str(pedDat)
```

## SNP genotypes

```{r}
inF <- file.path(projectDir, "simulations", "example_data", "genotype.txt")
stopifnot(file.exists(inF))
genos <- read.table(inF, header=FALSE, sep="\t", stringsAsFactors=FALSE)
rownames(genos) <- as.character(genos[,1])
genos <- as.matrix(genos[,-1])
stopifnot(all(genos >= 0 & genos <= 2))
str(genos)
genos[1:3, 1:6]
```

Allele frequencies:
```{r}
AFs <- colMeans(genos) / 2
hist(AFs)
```

## Lineages

```{r}
inF <- file.path(projectDir, "simulations", "example_data", "IBD.txt")
stopifnot(file.exists(inF))
IBDs <- read.table(inF, header=FALSE, sep="\t", stringsAsFactors=FALSE)
str(IBDs, list.len=6)
IBDs[1:3, 1:6]
```



# First goal: understand the genetic determinism of complex traits

## Expected IBD relationships from the pedigree

Reformat the input data into a specific object with the `kinship2` package:
```{r}
ped <- pedigree(id=pedDat$ind_id,
                dadid=pedDat$father_id,
                momid=pedDat$mother_id,
                sex=pedDat$sex)
ped
```

Plot:
```{r, fig.width=10}
if(FALSE){
  plot(ped)
}
```

Estimate the expected IBD:
```{r}
A <- 2 * kinship(ped)
dim(A)
```

Plot:
```{r, fig.width=10}
image(as(A, "Matrix"), main="A (pedigree)")
hist(diag(A))
hist(A[upper.tri(A)])
```

## Realized relationships assuming HWE and no DL

VanRaden P. 2008. Efficient methods to compute genomic predictions. Journal of Dairy Science. 91(11):4414–4423. doi:10.3168/jds.2007-0980.

```{r}
tmp <- matrix(rep(1, nrow(genos))) %*% (2 * AFs)
Z <- genos - tmp
rm(tmp)
GRM_vr <- tcrossprod(Z, Z) / (2 * sum(AFs * (1 - AFs)))
```

Plot:
```{r, fig.width=10}
image(as(GRM_vr, "Matrix"), main="GRM (Van Raden)")
hist(diag(GRM_vr))
hist(GRM_vr[upper.tri(GRM_vr)])
```

## True IBD relationships

```{r}

```

## Astle-Balding

Astle W, Balding D. 2009. Population structure and cryptic relatedness in genetic association studies. Statistical Science. 24(4):451–471. doi:10.1214/09-sts307.

## Yang

Yang J, Benyamin B, McEvoy B, Gordon S, Henders A, Nyholt D, Madden P, Heath A, Martin N, Montgomery G, et al. 2010. Common SNPs explain a large proportion of the heritability for human height. Nature genetics. 42(7):565–569. doi:10.1038/ng.608.

## LDAK

Speed D, Balding D. 2014. MultiBLUP: improved SNP-based prediction for complex traits. Genome Research. 24(9):1550–1557. doi:10.1101/gr.169375.113.

Speed D, Hemani G, Johnson M, Balding D. 2012. Improved heritability estimation from genome-wide SNPs. American journal of human genetics. 91(6):1011–1021. doi:10.1016/j.ajhg.2012.10.010.

```{r}

```


## Compare additive relationships

```{r}

```

## Estimate variance components

\begin{align}
\boldsymbol{y} = X \boldsymbol{\beta} + Z \boldsymbol{u} + \boldsymbol{\epsilon}
\end{align}

```{r}
hist(pedDat$phenotype)
plot(pedDat$breeding_value, pedDat$phenotype)

```


<!-- ==================================================================== -->

# Second goal: predict the breeding values of non-phenotyped individuals

```{r}

```


# Appendix

```{r}
print(sessionInfo(), locale=FALSE)
```
