---
title: "Genomic prediction for quantitative traits"
authors: "E. Tourrette, T. Flutre, B. Servin"
format:
  html:
    toc: true
    toc-depth: 5
    toc-location: left
    number-sections: true
editor: visual
---


# Preamble

Set the paths:
```{r}
projectName <- "QLife2025"
projectDir <- ""
if(Sys.info()["user"] == "tflutre"){
  projectDir <- file.path("~/Bureau/quick/", projectName)
} else if(Sys.info()["user"] == "eliset"){
  projectDir <- file.path("~/Desktop/qlife_2025")
}
projectDir <- path.expand(projectDir)
stopifnot(dir.exists(projectDir))
```

Load the dependencies:
```{r}
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(kinship2))
suppressPackageStartupMessages(library(sommer))
suppressPackageStartupMessages(library(rrBLUP))
```



# Main objective

In the goal of understanding the genetic determinism of complex traits, this session will show how to answer the question "how to replace pedigree-based kinship with SNP-based kinship?".




# Reminder of central concepts

Genetics is the study of heritable variation, and the goals of quantitative genetics are to (i) partition phenotypic variance into genetic vs. environmental components, (ii) predict genotypic values based on the resemblance between relatives, and (iii) find the underlying loci contributing to the genetic variance and values.
In the following, the most important concepts are briefly summarized for traits displaying quantitative variation; see Lynch and Walsh (1998) as well as [de Los Campos (2015)](https://dx.plos.org/10.1371/journal.pgen.1005048) from more details.

## Definition of the additive genetic variance

Let us imagine a population of $n$ diploid individuals at Hardy-Weinberg equilibrium (HWE), phenotyped for a given trait.
In the theory of quantitative genetics, if part of the phenotypic variation is of genetic origin, it means that (i) we assume the existence of $q$ quantitative trait locus (QTLs) with genetic effects on the trait, and (ii) the genetic variance comes from the individuals differing by their genotypes at these QTLs.
With $i \in \{1,\ldots,n\}$ to index the individuals, the phenotypic value of individual $i$ is noted $y_i$.
With $j \in \{1,\ldots,q\}$ to index the QTLs, the genotype of this individual at QTL $j$ corresponds to its allelic content and is noted $z_{ij}$.
The genotype of individual $i$ at all QTLs is the vector $\boldsymbol{z_i}$.
To simplify, we also center all variables: $E[y_i] = E[z_{ij}] = 0$.
The **genetic value** of an individual is then defined as its expected phenotypic value, over all possible environments, given its QTL genotypes:
\begin{align}
g_i = E_{env}[y_i|\boldsymbol{z_i}]
\end{align}

In a given macro-environment, an analysis of variance is applied with the phenotypic value as response variable and the genetic value as explanatory factor: $y_i = g_i + \epsilon_i$, where $\epsilon_i$ is a residual error corresponding to micro-environmental variation.
By construction: $\sigma_p^2 = \sigma_g^2 + \sigma^2$.
Moreover, the squared correlation between $g$ and $y$, known as the **broad-sense heritability**, equals: $H^2 := \sigma_g^2 / \sigma_p^2$.

The **genetic variance** for the trait in the population is informally defined as the variance of the genetic values.
However, these combine two sources of information, both being unknown: the genetic effects of the QTLs as well as the genotypes of the individual at the QTLs.
In the classical framework of quantitative genetics, inspired by the frequentist paradigm of statistics, the QTL genetic effects are considered as "fixed parameters" whereas the individual genotypes at QTLs are considered as "random variables".
As a result, the genetic variance is the expectation, over the individual genotypes at the QTLs, of the squared deviation of the genetic values from their population mean:
\begin{align}
\sigma_g^2 &:= E_z[ (g_i - E_z[g_i])^2 ] = E_z[g_i^2] - (E_z[g_i])^2 \; \text{(by definition)} \\
&= E_z[g_i^2] - (E_z[E_{env}[y_i|\boldsymbol{z_i}]])^2 = E_z[g_i^2] - (E_{env}[y_i])^2 \; \text{(law of total expectation)}  \\
&= E_z[g_i^2]
\end{align}

Most likely, the genetic value is a non-linear function of QTL genetic effects and genotypes.
Moreover, under sexual reproduction (on which we focus here), a key insight is that a parent only transmits to its offsprings a single allele at each QTL, not its whole genotype.
The strategy hence consists in approximating the genetic value by linearly regressing it on allele content: $g_i = \boldsymbol{z_i}^T \boldsymbol{\alpha} + \delta_i$ where $\boldsymbol{\alpha}$ is the vector of **additive genetic effects** of the QTLs (also known as allele substitutions) and $\delta_i$ is a residual deviation including intra-locus and inter-locus interactions (dominance and epistasis, respectively) as well as QTL-environment interactions.
The **additive genetic value** is then defined as: $g_{a,i} := \boldsymbol{z_i}^T \boldsymbol{\alpha}$.
Finding the best approximation, in the sense of the least squares, leads to: $\boldsymbol{\alpha} = Cov(z_i,z_i^T)^{-1} Cov(z_i,g_i)$ where $\Sigma_z := Cov(z_i,z_i^T)$ corresponds to the (co)variances of allelic content.

The **breeding value** (BV) of individual $i$ is defined as twice the expected deviation of the mean phenotype of this individual's progeny from the population mean when mated at random to other individuals from the population: $BV_i = 2 \times Exp[y_{\text{offsprings of $i$}} - \mu]$, so that the breeding value of an individual quantifies its value as a parent.
The breeding value is not exactly equal to the additive genetic value because it is not only half the additive genetic effects that are transmitted from parent to offspring, but also a small part of epistatic interactions.

Based on the linear regression above, the **additive genetic variance** is defined as: $\sigma_a^2 := \boldsymbol{\alpha}^T \Sigma_z \boldsymbol{\alpha}$.
More specifically:
\begin{align}
\sigma_a^2 = \sum_j Var(z_{ij}) \alpha_j^2 \; + \; 2 \sum_j \sum_{j' > j} Cov(z_{ij},z_{ij'}) \alpha_j \alpha_{j'}
\end{align}
The additive genetic variance thus not only depends on the contribution of genetic effects and variance of allelic contents to the genetic variance, but also on the covariance of allelic contents, also known as **linkage disequilibrium**.
The **narrow-sense heritability** is defined as: $h^2 := \sigma_a^2 / \sigma_p^2$.

## Genetic relationships without genomic markers

Two individuals are considered genetically identical for a given trait, that is, they have the same genetic value, if they have the same genotypes at the QTLs for that trait.
In other words, what counts for two individuals to be genetically identical is if they have alleles that are **identical by states** (IBS) at causal genes.
But these genes are usually unknown.
Still, related individuals are more likely to be genetically similar because they are more likely to have inherited the same alleles at the QTLs.
Two alleles will be **identical by descent** (IBD) if they come from the exact same allele in an ancestor some generations ago.

To quantify this, the **coancestry coefficient** between individuals $i$ and $i'$, noted $\phi_{a,ii'}$, is the probability that one allele taken at random from individual $i$ is IBD with an allele randomly drawn from individual $j$.
It can then be shown that the covariance between the additive genetic values of both individuals equals twice the product of the coancestry coefficient with the additive genetic variance: $Cov(g_{a,i},g_{a,i'}) = 2 \, \phi_{a,ii'} \, \sigma_a^2$.
The **additive genetic relationship** is twice the coancestry coefficient: $A_{ii'} := 2 \, \phi_{a,ii'}$.
When the causal genes are unknown (most of the cases), the resemblance between relatives can thus be used to estimate the additive genetic variance.

Concretely, this requires to phenotype $n$ individuals from multiple generations for which we know their **pedigree**.
The founders of the pedigree, assumed unrelated, form the **base population**: the additive genetic variance of interest is the variance of the founder's additive genetic values.
Efficient algorithms exist to compute coancestry coefficients, and all pairwise additive genetic relationships are gathered into the so-called numerator relationship matrix noted $A$.

## Estimation of the additive genetic variance without genomic markers

Estimating variance components requires a **linear mixed model** (LMM) whose generic form is: $\boldsymbol{y} = X \boldsymbol{\beta} + Z \boldsymbol{u} + \boldsymbol{\epsilon}$, where $\boldsymbol{y}$ is the $n$-vector of phenotypes, $X$ is the $n \times p$ design matrix for the explanatory factors whose effects, $\boldsymbol{\beta}$, are modeled as "fixed", $Z$ is the $n \times q$ design matrix for the explanatory factors whose effects, $\boldsymbol{u}$, are modeled as "random", and $\boldsymbol{\epsilon}$ are residual errors.
Moreover, $E[\boldsymbol{u}]=0$, $Var(\boldsymbol{u})=G$, $E[\boldsymbol{\epsilon}]=0$, $Var(\boldsymbol{\epsilon})=R$, and $Cov(\boldsymbol{u},\boldsymbol{\epsilon})=0$.

In classical quantitative genetics, the specific type of LMM used to estimate the additive genetic variance is called "variance-component" (vc-LMM).
The additive genetic values are modeled as random and their variance-covariance matrix is $\sigma_a^2 \, A$.
Moreover, usually $R = \sigma^2 \, \text{Id}_n$.
In practice, the variance components are estimated by **restricted maximum likelihood** (ReML).

## Prediction of additive genetic values without genomic markers

The **mixed model equations** (MMEs) are used to obtain formulas for the best linear unbiased estimator (BLUE) of the fixed effects, and the **best linear unbiased predictor** (BLUP) of the random effects: $BLUP(\boldsymbol{u}) = G Z^T V^{-1} (\boldsymbol{y} - X \boldsymbol{\beta})$ where $V :=  Z G Z^T + R$.
When the estimates of the variance components are used, the so-called empirical BLUPs (eBLUPs) can be computed.

The additive genetic values of unphenotyped individuals can be predicted as long as their additive genetic relationships with phenotyped individuals are known (pedigree).

TODO: add formula

This procedure is very useful but its accuracy can be much increased.
Indeed, unphenotyped individuals with the same genetic relationships get the same genetic values, as full siblings.
This is because the genetic relationships estimated from the pedigree are *expected* relationships assuming an infinite number of loci.
However, a parent passes a *sampled* half of its alleles to its offsprings, that is, a gamete carries a *random assortment* (RA) of parental alleles, also called **Mendelian deviations**.
The breeding value of an offspring resulting from the cross of individuals $i$ and $i'$ can hence be modeled as: $BV_o = (1/2) BV_i + RA_i + (1/2) BV_{i'} + RA_{i'}$.

Importantly, Mendelian deviations are what breeding relies on in order to obtain a positive response to selection.
It is thus important to take these deviations into account, and which is made possible by genotyping individuals at $L$ markers.
However, the usage of genomic markers in this purpose raises various questions, that will be explored below with simulations.

<!--
The definition of the substitution effect highlights the important fact that a gene does not "make" a trait, rather it is allelic differences that cause phenotypic differences ([Orgogozo et al, 2015](http://www.frontiersin.org/Evolutionary_and_Population_Genetics/10.3389/fgene.2015.00179/abstract); [de Vienne, 2021](https://link.springer.com/10.1007/s10709-021-00134-6)).
That is why the first goal, in this framework, consists in an analysis of variance ([Lewontin, 2006](http://dx.doi.org/10.1093/ije/dyl062)).
Moreover, the fact that the substitution effect depends on the allele frequency means that, as the latter changes due to evolutionary forces (drift, selection, migration), the former will change as a result.

To introduce what genetic effects are, let us assume first that there is a single biallelic QTL, with alleles noted $A_1$ and $A_2$, the reference allele being arbitrarily chosen as being $A_2$, the frequency of the reference allele noted $f$, and the **gene content**, $N_k$, being the number of copies of allele $k$ for a given genotype.
From a functional or physiological point of view, the genotypic value $G_{kl}$, i.e., the genetic value of genotype $A_k A_l$, is defined by its additive **gene action**, $a := (G_{22} - G_{11}) / 2$, with $G_{11}=-a$ and $G_{22}=+a$, and its dominance gene action, $d := G_{12} - (G_{11}+G_{22})/2$.
The linear regression of genotypic value on gene contents hence is: $G_{kl} = \mu + \alpha_1 N_1 + \alpha_2 N_2 + \delta_{kl}$, where the $\alpha_k$'s are called **additive effects**.
It can be reformulated into: $G_{kl} = \mu + \alpha N_2 + \delta_{kl}$, where the slope is called the **substitution effect**: $\alpha := \alpha_2 - \alpha_1 = a + (1 - 2 f) d$ under HWE.

The part of the genotypic value corresponding to the sum of additive effects is called the **additive genetic value**: $G_{A} := \alpha_k + \alpha_l$ in the case of a single QTL.
The **additive genetic variance** is then defined as the variance of the additive genetic values: $\sigma_A^2 := 2 f (1 - f) \alpha^2$ in the case of a single QTL under HWE.
Note that this variance is population-specific since it depends on the allele frequency and, because it depends on $\alpha$ that itself depends on both $a$ and $d$, the relative magnitude of genetic variances does not reflect the relative importance of gene actions ([Huang and Mackay, 2016](http://dx.plos.org/10.1371/journal.pgen.1006421)).
The **narrow-sense heritability** is the proportion of phenotypic variance that corresponds to the additive genetic variance: $h^2 := \sigma_A^2 / \sigma_P^2$.
It plays a key role because the response to selection after one generation equals $h^2 \, S$ where $S$ is the selection differential, this relation being known as the **breeder's equation**.
-->


<!--
In this part we are interested in decomposing the overall phenotypic variance in several components, notably the genetic variance and, more specifically, the additive genetic variance.
How can it be estimated, and which improvements do markers provide?
This part is heavily inspired by [Toro et al (2011)](http://www.gsejournal.org/content/43/1/27), [Legarra (2016)](http://dx.doi.org/10.1016/j.tpb.2015.08.005) and [Schreck et al (2019)](https://academic.oup.com/genetics/article/213/2/379/5930631).

TODO: explain really what $\sigma_u^2$ is, as trace of cov(ui,uj) (see Schreck et al 2019)
-->



# Run the simulation

In this tutorial, we will play with two simulate data sets.

TODO: describe them

```{r}
## TODO
```



# Load all the data

## Pedigree

```{r}
inF <- file.path(projectDir, "simulations", "example_data", "pedigree.txt")
stopifnot(file.exists(inF))
pedDat <- read.table(inF, header=TRUE, sep=" ", stringsAsFactors=FALSE)
str(pedDat)
length(unique(pedDat$ind_id))
table(pedDat$generation)
```

## SNP genotypes

```{r}
inF <- file.path(projectDir, "simulations", "example_data", "genotype.txt")
stopifnot(file.exists(inF))
genos <- read.table(inF, header=TRUE, sep="\t", stringsAsFactors=FALSE,
                    row.names=1, check.names=FALSE)
genos <- as.matrix(genos)
stopifnot(all(genos >= 0 & genos <= 2))
str(genos)
genos[1:3, 1:6]
```

## SNP effects

```{r}
inF <- file.path(projectDir, "simulations", "example_data", "SNP_INFO.txt")
stopifnot(file.exists(inF))
snpInfos <- read.table(inF, header=TRUE, sep="\t")
str(snpInfos)
head(snpInfos)
length(unique(snpInfos$snp_id))
table(snpInfos$chr_id)
```

## Lineages

```{r}
inF <- file.path(projectDir, "simulations", "example_data", "IBD.txt")
stopifnot(file.exists(inF))
IBDs <- read.table(inF, header=TRUE, sep="\t", stringsAsFactors=FALSE, check.names=FALSE)
str(IBDs, list.len=6)
IBDs[1:3, 1:6]
```




# Estimating relationships

## True IBD relationships

TODO Elise

```{r}

```

## Expected genetic relationships from the pedigree

Reformat the input data into a specific object with the `kinship2` package:
```{r}
ped <- pedigree(id=pedDat$ind_id,
                dadid=pedDat$father_id,
                momid=pedDat$mother_id,
                sex=pedDat$sex)
ped
```

Plot:
```{r, fig.width=10}
if(FALSE){
  plot(ped)
}
```

Estimate the coancestry coefficient, the `kinship` function using a recursive algorithm described in Lange (1997):
```{r}
A <- 2 * kinship(ped)
dim(A)
```

Plot:
```{r, fig.width=10}
image(as(A, "Matrix"), main="A (pedigree)")
hist(diag(A))
hist(A[upper.tri(A)])
```

## Realized relationships

Let the markers be indexed by $l \in \{1,\ldots,L\}$, $f_l$ being the allele frequency of marker $l$ and $M_{il}$ being the number of reference allele for individual $i$ at marker $l$, so that $M$ is the $n \times L$ matrix of SNP genotypes.
[Van Raden (2008)](http://dx.doi.org/10.3168/jds.2007-0980) proposed two estimators, relying on the molecular covariance and the allele frequency at each locus.
Both estimators assume that the reference population is a population at HWE with allele frequencies equal to the allele frequencies observed in the sample.

### First estimator

The estimator of molecular covariance is computed for each locus, summed over all locus, and standardized by the total molecular variance:
\begin{align}
\hat{\phi}_{a,ii'}^{(VR1)} = \frac{\sum_{l=1}^L (M_{il} - 2 \hat{f}_l)(M_{i'l} - 2 \hat{f}_l)}{2 \sum_{l=1}^L \hat{f}_l (1 - \hat{f}_l)}
\end{align}

In matrix form, the matrix of SNP genotypes, $M$, is centered, $Z$, so that $GRM_{VR1} := Z Z^T / (2 \sum_l \hat{f}_l (1 - \hat{f}_l))$.

```{r}
estimVanRanden1 <- function(genos, AFs){
  tmp <- matrix(rep(1, nrow(genos))) %*% (2 * AFs)
  Z <- genos - tmp
  tcrossprod(Z, Z) / (2 * sum(AFs * (1 - AFs)))
}
```

#### Generation 0

If the sample is at HWE, the diagonal should be centered on 1 and the off-diagonal on 0:
```{r}
inds_gen0 <- pedDat$ind_id[pedDat$generation == 0]
AFs_gen0 <- colMeans(genos[inds_gen0,]) / 2
hist(AFs_gen0)
GRM_VR1_gen0 <- estimVanRanden1(genos[inds_gen0,], AFs_gen0)
image(as(GRM_VR1_gen0, "Matrix"), main="GRM (Van Raden 1, generation 0)")
hist(diag(GRM_VR1_gen0), main="diag of GRM (Van Raden 1, generation 0)")
hist(GRM_VR1_gen0[upper.tri(GRM_VR1_gen0)], main="offdiag of GRM (Van Raden 1, generation 0)")
```

#### All generations

```{r, fig.width=10}
AFs <- colMeans(genos) / 2
hist(AFs)
GRM_VR1 <- estimVanRanden1(genos, AFs)
image(as(GRM_VR1, "Matrix"), main="GRM (Van Raden 1, all generations)")
hist(diag(GRM_VR1), main="diag GRM (Van Raden 1, all generations)")
hist(GRM_VR1[upper.tri(GRM_VR1)], main="offdiag GRM (Van Raden 1, all generations)")
```

### Second estimator

The estimator of molecular covariance is computed for each locus, standardized, and summed over all locus:
\begin{align}
\hat{\phi}_{a,ii'}^{(VR2)} = \sum_{l=1}^L \frac{(M_{il} - 2 f_l)(M_{i'l} - 2 f_l)}{2 f_l (1 - f_l)}
\end{align}

```{r}
## TODO
```

### Other estimators

[Yang et al (2010)](http://dx.doi.org/10.1038/ng.608) proposed an estimator similar to the second one from Van Raden (2008).

<!--
Astle W, Balding D. 2009. Population structure and cryptic relatedness in genetic association studies. Statistical Science. 24(4):451–471. doi:10.1214/09-sts307.
-->

## LDAK estimator

Speed D, Balding D. 2014. MultiBLUP: improved SNP-based prediction for complex traits. Genome Research. 24(9):1550–1557. doi:10.1101/gr.169375.113.

Speed D, Hemani G, Johnson M, Balding D. 2012. Improved heritability estimation from genome-wide SNPs. American journal of human genetics. 91(6):1011–1021. doi:10.1016/j.ajhg.2012.10.010.

```{r}
## TODO
```

## Estimator from Schreck et al

Schreck N, Piepho H-P, Schlather M. 2019. Best Prediction of the Additive Genomic Variance in Random-Effects Models. Genetics. 213(2):379–394. doi:10.1534/genetics.119.302324.

```{r}
## TODO
```

## Compare additive relationships

```{r}

```

## Estimate variance components

```{r}
hist(pedDat$pheno1)
plot(pedDat$bv1, pedDat$pheno1)
```


<!-- ==================================================================== -->

# Predict the breeding values of non-phenotyped individuals

TODO: compare estimates of SNP effects ($\hat{\beta}$) with true effects

```{r}

```


# Appendix

```{r}
print(sessionInfo(), locale=FALSE)
```
